src/compiler/factory.ts

```diff
@@ -1618,17 +1618,38 @@ namespace ts {
         );
     }
 
-    function createReactNamespace(reactNamespace: string, parent: JsxOpeningLikeElement) {
+    export function createEntityName(entityName: string) {

```

Comment:
Perhaps a `parseIsolatedEntityName` could set a flag in the parser to have it treat the source as synthetic.